Almost finish the `env` module

tbu- opened the pull request:
The only function missing is `home_dir`

tbu- committed:
Almost finish the `env` module

Only function missing is `home_dir`

homunkulus commented:
:umbrella: The latest upstream changes (presumably #110) made this pull request unmergeable. Please resolve the merge conflicts.

tbu- committed:
Update the remaining parts to Rust 1.16.0

tbu- commented:
@homunkulus try

homunkulus commented:
:hourglass: Trying commit a2c6ab8 with merge 8034123...

tbu- commented:
Hm. Travis doesn't appear here, but this shows as passed on Travis's website.

homunkulus commented:
:sunny: Test successful - [status-travis](https://travis-ci.org/japaric/steed/builds/224805410)
State: approved= try=True

japaric commented on src/libc/mod.rs:
```diff
@@ -80,6 +83,86 @@ pub unsafe fn strlen(cs: *const c_char) -> size_t {
+    pub value: Cow<'static, [u8]>,
+}
+
+#[allow(unused)]
```
Hmm, why is the `allow(unused)` required? `ENV_PTR` is used in `_init` down below, no? (Same with AUXVAL)

japaric commented on src/libc/mod.rs:
```diff
@@ -80,6 +83,86 @@ pub unsafe fn strlen(cs: *const c_char) -> size_t {
+    if input.is_empty() {
+        return None;
+    }
+    let pos = memchr(b'=', &input[1..]).map(|p| p + 1);
```
It wasn't obvious to me why `input` is sliced like this. Could you copy this comment from `std`?

```
        // Strategy (copied from glibc): Variable name and value are separated
        // by an ASCII equals sign '='. Since a variable name must not be
        // empty, allow variable names starting with an equals sign. Skip all
        // malformed lines.
```

I'm not sure if it's around already but if it is then all these `1..` slicing operations should point out to that comment.

japaric commented on src/sys/linux/os.rs:
```diff
@@ -1,5 +1,25 @@
+use vec;
+
+static ENV_LOCK: () = ();
+// TODO(steed): Synchronize environment access once we have mutexes.
```
Could you open an issue about this? Now that we have threads this one may go very wrong if two or more threads try to `set_env` the same variable :-/.

tbu- replied:
#143

japaric commented on src/libc/mod.rs:
```diff
@@ -80,6 +83,86 @@ pub unsafe fn strlen(cs: *const c_char) -> size_t {
+    &*ENV
+}
+
+pub unsafe fn environ() -> *const *const c_char {
```
Hmm, does this function actually make it to the `std` public API? It any case, I think this should panic like `libc::errno`. We should probably document these differences from `std` since there now would be two of them.

tbu- replied:
No, it doesn't make it to the public API, this is the stub for spawning new commands right now.

japaric replied:
Hmm, does that means that spawned processes don't inherit their parent's environment? We should create an issue about that if that's the case.

japaric reviewed (commented):
Excellent work as usual @tbu- 👍 

r=me with nits addressed.

@homunkulus delegate+

japaric commented:
@homunkulus delegate+

(I think the review comment is not picked by Homu)

homunkulus commented:
:v: @tbu- can now approve this pull request

tbu- committed:
Add a tracking issue, add a comment, remove an annotation

tbu- commented:
Since `environ` is *not* part of the public API, and since it was stubbed before this pull request, (I just moved it to `libc`), I'm going to merge this. I hope this is OK.

Addressed the other comments.

@homunkulus r+

homunkulus commented:
:pushpin: Commit 3c2bd06 has been approved by `tbu-`

<!-- @homunkulus r=tbu- 3c2bd063f06a6696c7e3f0ab7e0faace8250e20d -->

homunkulus commented:
:hourglass: Testing commit 3c2bd06 with merge db8fb00...

homunkulus commented:
:sunny: Test successful - [status-travis](https://travis-ci.org/japaric/steed/builds/224855768)
Approved by: tbu-
Pushing db8fb00cdc7526fa431c95587e131bfa7baf3a55 to master...
